Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/BuildTarget/Common/TestImpactBuildGraph.h

```diff
@@ -0,0 +1,174 @@
+                            "Couldn't find build dependency '%s' for build target '%s'",
+                            buildDependency.c_str(),
+                            buildTarget.GetTarget()->GetName().c_str());
+                        continue;
+                    }
+
+                    const auto* buildDependencyVertex = addOrRetrieveVertex(buildDependencyTarget.value());
+                    resolveDependencies.insert(buildDependencyVertex);
+                }
+            };
+
+            auto& vertex = *addOrRetrieveVertex(buildTarget);
+            resolveDependencies(buildTarget.GetTarget()->GetDependencies().m_build, vertex.m_dependencies.m_build);
+            resolveDependencies(buildTarget.GetTarget()->GetDependencies().m_runtime, vertex.m_dependencies.m_runtime);
+        }
+
+        // Build depender graphs
+        for (auto& [buildTarget, vertex] : m_buildGraphVertices)
+        {
+            for (auto* dependency : vertex.m_dependencies.m_build)
+            {
+                auto& dependencyVertex = m_buildGraphVertices.find(dependency->m_buildTarget)->second;
+                dependencyVertex.m_dependers.m_build.insert(&vertex);
+            }
+
+            for (auto* dependency : vertex.m_dependencies.m_runtime)
+            {
+                auto& dependencyVertex = m_buildGraphVertices.find(dependency->m_buildTarget)->second;
+                dependencyVertex.m_dependers.m_runtime.insert(&vertex);
+            }
+        }
+    }
+
+    template<typename ProductionTarget, typename TestTarget>
+    const BuildGraphVertex<ProductionTarget, TestTarget>* BuildGraph<ProductionTarget, TestTarget>::GetVertex(
+        const BuildTarget<ProductionTarget, TestTarget>& buildTarget) const
+    {
+        if (const auto it = m_buildGraphVertices.find(buildTarget) != m_buildGraphVertices.end())
+        {
+            return &it->second;
+        }
+
+        return nullptr;
+    }
+
+    template<typename ProductionTarget, typename TestTarget>
+    const BuildGraphVertex<ProductionTarget, TestTarget>& BuildGraph<ProductionTarget, TestTarget>::
+        GetVertexOrThrow(
+        const BuildTarget<ProductionTarget, TestTarget>& buildTarget) const
+    {
+        const auto vertex = GetVertex(buildTarget);
+        TestImpact_Eval(
+            vertex, 
+            BuildTargetException, 
+            AZStd::string::format().c_str("Couldn't find build target '%s'", 
+            buildTarget.GetTarget()->GetName().c_str()));
+        return vertex;
+    }
+} // namespace TestImpact
+
+namespace AZStd
+{
+    //! Hash function for BuildTarget types for use in maps and sets.
+    template<typename ProductionTarget, typename TestTarget>
+    struct hash<TestImpact::BuildGraphVertex<ProductionTarget, TestTarget>>
+    {
+        size_t operator()(const TestImpact::BuildGraphVertex<ProductionTarget, TestTarget>& vertex) const noexcept
+        {
+            return reinterpret_cast<size_t>(vertex.m_buildTarget.GetTarget());
```

Comment:
question: Does this need to be a `reinterpret_cast`? Would a `static_cast` suffice?

Reply:
It's to get the numerical value of the underlying object address, so `static_cast` will not work.